flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala

```diff
@@ -1463,21 +1465,23 @@ class CodeGenerator(
     */
   def addReusableFunction(function: UserDefinedFunction): String = {
     val classQualifier = function.getClass.getCanonicalName
-    val fieldTerm = s"function_${classQualifier.replace('.', '$')}"
+    val functionSerializedData = serialize(function)
+    val fieldTerm =
+      s"""
+         |function_${classQualifier.replace('.', '$')}_${DigestUtils.md5Hex(functionSerializedData)}
```

Comment:
I think it makes sense to keep the same implementation for UDF and UDTF. What about code generate a byte[] field which hold the serialized function (proposed by Fabian)? 

>I see two options:
>1. make the UDF a member of wrapping function. It might be a bit tricky to pass the reference into the code-gen'd function.
>2. add a final byte[] field into the code-gen'd function that holds the serialized UDF object and deserialize during initialization. This will blow up the code-gen'd string but might work well.